blog/content/post/2024-08-19-secureboot-freebsd.md

```diff
@@ -0,0 +1,397 @@
+```
+
+Add a safety factor of a couple hundred bytes (~512) to this number, and record
+it.
+
+### Build the Loader With Extra Space
+
+This step requires you to have the source code for your system. You might have
+checked the `src` component during OS installation, then the `/usr/src/`
+directory should be already populated. If not, clone the source code from
+GitHub, substituting `14.1` with the release you are currently using:
+
+```bash
+git clone -b releng/14.1 https://git.freebsd.org/src.git /usr/src
+```
+
+Now, recall your bootfs size and substitute `${BOOTFS_SIZE_PLUS_SAFETY}` below:
+
+```bash
+cd /usr/src/stand
+make MD_IMAGE_SIZE=${BOOTFS_SIZE_PLUS_SAFETY}
+```
+
+### Embed bootfs image in the loader.efi
+
+The build results will be available under `/usr/obj`.
+
+* If you have chosen to use the legacy FORTH loader, the path to the file that
+  interests you should be similar to
+  `/usr/obj/usr/src/amd64.amd64/stand/efi/loader_4th/loader_4th.efi`.
+
+* If you have chosen the Lua loader, the path should be similar to
+  `/usr/obj/usr/src/amd64.amd64/stand/efi/loader_lua/loader_lua.efi`
+
+Copy the appropriate file and use the available embedding utility:
+
+```bash
+cp /usr/obj/${PATH_TO_LOADER}/loader.efi ~/
+/usr/src/sys/tools/embed_mfs.sh ~/loader.efi ~/bootfs.img
+```
+
+You might need to make `embed_mfs.sh` executable:
+
+```bash
+chmod +x /usr/src/sys/tools/embed_mfs.sh
+```
+
+At this point, `loader.efi` is an UEFI-bootable binary, consisting of the
+FreeBSD bootloader and kernel. The last remaining step to Secure Boot
+compatibility is generating keys and signing the binary.
+
+## Signing the binary
+
+FreeBSD includes a tool for signing EFI executables - `uefisign`. There is also
+a utility provided to generate example keys and certificates needed to sign an
+executable.
+
+You can generate a self-signed certificate and use it to sign a binary as
+follows:
+
+```bash
+/usr/share/examples/uefisign/uefikeys testcert
+uefisign -c testcert.pem -k testcert.key -o signed-loader.efi loader.efi
+```
+
+As earlier, make sure the script is marked as executable.
+
+You should now have a `signed-loader.efi` and a `testkey.cer` file. The loader
+file is what we're going to be booting from Dasharo, and the `.cer` file is our
+custom certificate we need to enroll, so that Secure Boot can verify the
+loader's signature against it.
+
+## Testing in QEMU
+
+To spare yourself the trouble of recovering a broken OS installation, it is
+recommended to test the binary in an emulated environment.
+
+Make sure you have an up-to-date installation of QEMU on your system, and get
+the latest QEMU release of Dasharo
+[here](https://github.com/Dasharo/edk2/releases).
+
+### Launching QEMU with an emulated filesystem
+
+Create a directory for the QEMU firmware files, and a directory for the test
+EFI files:
+
+```bash
+mkdir -p ~/qemu_test/efi
+```
+
+Download `OVMF_CODE_RELEASE.fd` and `OVMF_VARS_RELEASE.fd` and place them in
```

Comment:
From where user should download it?

Reply:
[Added link](https://github.com/3mdeb/news-and-ideas/pull/603/commits/469cc84541acf2428e46af01de19cf63b5bb0466#diff-1e142bc7519e97361fc0d3867abcb7fa3d257a4b5ccb79795e5e00acebf7d22fR244)